File: app/src/test/java/com/thoughtworks/pafsilva/testautomationworkshop/login/LoginPresenterImplTest.java
package com.thoughtworks.pafsilva.testautomationworkshop.login;

import com.thoughtworks.pafsilva.testautomationworkshop.R;
import com.thoughtworks.pafsilva.testautomationworkshop.login.abstractions.LoginView;
import com.thoughtworks.pafsilva.testautomationworkshop.login.services.LoginService;
import com.thoughtworks.pafsilva.testautomationworkshop.login.services.UserLoginCallback;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import java.util.regex.Matcher;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class LoginPresenterImplTest {

    private LoginPresenterImpl loginPresenter;

    @Mock
    private LoginView loginView;

    @Mock
    private LoginService loginService;

    @Before
    public void setUp() throws Exception {
        loginPresenter = new LoginPresenterImpl(loginView, loginService);
    }

    @Test
    public void doLoginWhenClickOnLoginButton() throws Exception {
        loginPresenter.handleClick(R.id.btnLogin);

        verify(loginView).doLogin();
    }

    @Test
    public void callExternalBrowserWhenClickOnForgotPasswordButton() throws Exception {
        loginPresenter.handleClick(R.id.txtbtnForgotPassword);

        verify(loginView).navigateToBrowser();
    }

    @Test
    public void callValidationErrorWhenEmailIsIncorrect() throws Exception {
        loginPresenter.doLogin("incorrect", "password", false);

        verify(loginView).setEmailEditTextError("Please fill your email correctly");
    }

    @Test
    public void callValidationErrorWhenEmailIsEmpty() throws Exception {
        loginPresenter.doLogin("", "password", false);

        verify(loginView).setEmailEditTextError("This field is empty");
    }

    @Test
    public void callValidationErrorWhenPasswordIsIncorrect() throws Exception {
        loginPresenter.doLogin("test@test.com", "password", true);

        verify(loginView).setPasswordEditTextError("Your password must have at least 8 characters with letters and numbers");
    }

    @Test
    public void callValidationErrorWhenPasswordlIsEmpty() throws Exception {
        loginPresenter.doLogin("test@test.com", "", true);

        verify(loginView).setPasswordEditTextError("This field is empty");
    }

    @Test
    public void showsLoadingWhenEmailAndPasswordAreCorrect() throws Exception {
        loginPresenter.doLogin("test@test.com", "passwordWith1234567Numbers", true);

        verify(loginView).showLoading();
    }

    @Test
    public void callServiceWhenEmailAndPasswordAreCorrect() throws Exception {
        loginPresenter.doLogin("test@test.com", "passwordWith1234567Numbers", true);

        verify(loginService).callAuthenticationService(any(UserLoginCallback.class),
                eq("test@test.com"), eq("passwordWith1234567Numbers"));
    }
}

File: app/src/test/java/com/thoughtworks/pafsilva/testautomationworkshop/login/LoginPresenterImplTest.kt
package com.thoughtworks.pafsilva.testautomationworkshop.login

import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.eq
import com.thoughtworks.pafsilva.testautomationworkshop.R
import com.thoughtworks.pafsilva.testautomationworkshop.login.abstractions.LoginView
import com.thoughtworks.pafsilva.testautomationworkshop.login.services.LoginService
import com.thoughtworks.pafsilva.testautomationworkshop.login.services.UserLoginCallback
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class LoginPresenterImplTest {

    private lateinit var loginPresenter: LoginPresenterImpl

    @Mock
    private lateinit var loginView: LoginView

    @Mock
    private lateinit var loginService: LoginService

    @Before
    fun setUp() {
        loginPresenter = LoginPresenterImpl(loginView, loginService)
    }

    @Test
    fun doLoginWhenClickOnLoginButton() {
        loginPresenter.handleClick(R.id.btnLogin)

        verify(loginView).doLogin()
    }

    @Test
    fun callExternalBrowserWhenClickOnForgotPasswordButton() {
        loginPresenter.handleClick(R.id.txtbtnForgotPassword)

        verify(loginView).navigateToBrowser()
    }

    @Test
    fun callValidationErrorWhenEmailIsIncorrect() {
        loginPresenter.doLogin("incorrect", "password", false)

        verify(loginView).setEmailEditTextError("Please fill your email correctly")
    }

    @Test
    fun callValidationErrorWhenEmailIsEmpty() {
        loginPresenter.doLogin("", "password", false)

        verify(loginView).setEmailEditTextError("This field is empty")
    }

    @Test
    fun callValidationErrorWhenPasswordIsIncorrect() {
        loginPresenter.doLogin("test@test.com", "password", true)

        verify(loginView).setPasswordEditTextError("Your password must have at least 8 characters with letters and numbers")
    }

    @Test
    fun callValidationErrorWhenPasswordIsEmpty() {
        loginPresenter.doLogin("test@test.com", "", true)

        verify(loginView).setPasswordEditTextError("This field is empty")
    }

    @Test
    fun showsLoadingWhenEmailAndPasswordAreCorrect() {
        loginPresenter.doLogin("test@test.com", "passwordWith1234567Numbers", true)

        verify(loginView).showLoading()
    }

    @Test
    fun callServiceWhenEmailAndPasswordAreCorrect() {
        loginPresenter.doLogin("test@test.com", "passwordWith1234567Numbers", true)

        verify(loginService).callAuthenticationService(
            any<UserLoginCallback>(),
            eq("test@test.com"),
            eq("passwordWith1234567Numbers")
        )
    }
}